File: java/src/main/java/benchmark/model/v2/Message.java
package benchmark.model.v2;

import java.io.Serializable;
import java.util.Objects;

/** Single-level mixed-primitive record (type_id=message). */
public final class Message implements Serializable {
  private static final long serialVersionUID = 1L;

  public boolean fBool;
  public int fInt32;
  public long fInt64;
  public double fFloat64;
  public String fString;
  public boolean fBool2;
  public int fInt32_2;
  public String fString2;

  public Message() {}

  public Message(
      boolean fBool,
      int fInt32,
      long fInt64,
      double fFloat64,
      String fString,
      boolean fBool2,
      int fInt32_2,
      String fString2) {
    this.fBool = fBool;
    this.fInt32 = fInt32;
    this.fInt64 = fInt64;
    this.fFloat64 = fFloat64;
    this.fString = fString;
    this.fBool2 = fBool2;
    this.fInt32_2 = fInt32_2;
    this.fString2 = fString2;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof Message m)) return false;
    return fBool == m.fBool
        && fInt32 == m.fInt32
        && fInt64 == m.fInt64
        && Math.abs(m.fFloat64 - fFloat64) <= 1e-9
        && fBool2 == m.fBool2
        && fInt32_2 == m.fInt32_2
        && Objects.equals(fString, m.fString)
        && Objects.equals(fString2, m.fString2);
  }

  @Override
  public int hashCode() {
    return Objects.hash(fBool, fInt32, fInt64, fFloat64, fString, fBool2, fInt32_2, fString2);
  }
}

File: kotlin/src/main/kotlin/benchmark/model/v2/Message.kt
package benchmark.model.v2

import com.squareup.moshi.JsonClass
import kotlinx.serialization.Serializable
import kotlinx.serialization.protobuf.ProtoNumber
import java.io.Serializable as JavaSerializable

/** Single-level mixed-primitive record (type_id=message). */
@Serializable
@JsonClass(generateAdapter = true)
data class Message(
    @ProtoNumber(1) @JvmField var fBool: Boolean = false,
    @ProtoNumber(2) @JvmField var fInt32: Int = 0,
    @ProtoNumber(3) @JvmField var fInt64: Long = 0L,
    @ProtoNumber(4) @JvmField var fFloat64: Double = 0.0,
    @ProtoNumber(5) @JvmField var fString: String = "",
    @ProtoNumber(6) @JvmField var fBool2: Boolean = false,
    @ProtoNumber(7) @JvmField var fInt32_2: Int = 0,
    @ProtoNumber(8) @JvmField var fString2: String = "",
) : JavaSerializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Message) return false
        return fBool == other.fBool &&
            fInt32 == other.fInt32 &&
            fInt64 == other.fInt64 &&
            kotlin.math.abs(other.fFloat64 - fFloat64) <= 1e-5 &&
            fBool2 == other.fBool2 &&
            fInt32_2 == other.fInt32_2 &&
            fString == other.fString &&
            fString2 == other.fString2
    }

    override fun hashCode(): Int {
        var result = fBool.hashCode()
        result = 31 * result + fInt32
        result = 31 * result + fInt64.hashCode()
        result = 31 * result + fFloat64.hashCode()
        result = 31 * result + fString.hashCode()
        result = 31 * result + fBool2.hashCode()
        result = 31 * result + fInt32_2
        result = 31 * result + fString2.hashCode()
        return result
    }
}
